funsor/contract.py

```diff
@@ -0,0 +1,242 @@
+    meas_vars = frozenset(lhs.inputs)
+    int_vars = frozenset(rhs.inputs)
+    assert reduced_vars <= meas_vars | int_vars
+    progress = False
+    if not reduced_vars <= meas_vars:
+        rhs = rhs.reduce(ops.add, reduced_vars - meas_vars)
+        reduced_vars = reduced_vars & meas_vars
+        progress = True
+    if not reduced_vars <= int_vars:
+        lhs = lhs.reduce(ops.add, reduced_vars - int_vars)
+        reduced_vars = reduced_vars & int_vars
+        progress = True
+
+    if progress:
+        return Contract(lhs, rhs, reduced_vars)
+
+    return None
+
+
+class Contract(Funsor):
+
+    def __init__(self, lhs, rhs, reduced_vars):
+        assert isinstance(lhs, Funsor)
+        assert isinstance(rhs, Funsor)
+        assert isinstance(reduced_vars, frozenset)
+        inputs = OrderedDict([(k, d) for t in (lhs, rhs)
+                              for k, d in t.inputs.items() if k not in reduced_vars])
+        output = rhs.output
+        super(Contract, self).__init__(inputs, output)
+        self.lhs = lhs
+        self.rhs = rhs
+        self.reduced_vars = reduced_vars
+
+    def eager_subs(self, subs):
+        # basically copied from Reduce.eager_subs
+        subs = tuple((k, v) for k, v in subs if k not in self.reduced_vars)
+        if not any(k in self.inputs for k, v in subs):
+            return self
+        if not all(self.reduced_vars.isdisjoint(v.inputs) for k, v in subs):
+            raise NotImplementedError('TODO alpha-convert to avoid conflict')
+        return Contract(self.lhs.eager_subs(subs), self.rhs.eager_subs(subs),
+                        self.reduced_vars)
+
+
+@optimize.register(Contract, ATOMS[1:], ATOMS[1:], frozenset)
+@eager.register(Contract, ATOMS[1:], ATOMS[1:], frozenset)
+def contract_ground_ground(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    return (lhs * rhs).reduce(ops.add, reduced_vars)
+
+
+@eager.register(Contract, Tensor, Tensor, frozenset)
+def eager_contract_tensor_tensor(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    if reduced_vars:
+        out_vars = (frozenset(lhs.inputs) | frozenset(rhs.inputs)) - reduced_vars
+    else:
+        out_vars = {}
+
+    out_inputs = OrderedDict([
+        (v, lhs.inputs[v] if v in lhs.inputs else rhs.inputs[v]) for v in out_vars])
+
+    return Tensor(
+        opt_einsum.contract(lhs.data, list(lhs.inputs.keys()),
+                            rhs.data, list(rhs.inputs.keys()),
+                            list(out_vars), backend="torch"),
+        out_inputs
+    )
+
+
+@optimize.register(Contract, Funsor, Reduce, frozenset)
+def contract_reduce(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    # XXX should we be doing this conversion at all given that Reduce
```

Comment:
The use of `_make_base_lhs()` here seems weird to me, I'll have to think about it...

Reply:
I removed this rule